mempool/clist_sidecar.go

```diff
@@ -0,0 +1,655 @@
+}
+
+var _ PriorityTxSidecar = &CListPriorityTxSidecar{}
+
+type Key struct {
+	height, bundleID int64
+}
+
+// NewCListSidecar returns a new sidecar with the given configuration
+// takes in the logger for the mempool
+func NewCListSidecar(
+	height int64,
+	memLogger log.Logger,
+	mevMetrics *mev.Metrics,
+) *CListPriorityTxSidecar {
+	sidecar := &CListPriorityTxSidecar{
+		txs:                    clist.New(),
+		height:                 height,
+		heightForFiringAuction: height + 1,
+		logger:                 memLogger,
+		metrics:                mevMetrics,
+	}
+	sidecar.cache = NewLRUTxCache(10000)
+	return sidecar
+}
+
+func (sc *CListPriorityTxSidecar) PrettyPrintBundles() {
+	fmt.Println("-------------")
+	for bundleIDIter := 0; bundleIDIter <= int(sc.maxBundleID); bundleIDIter++ {
+		bundleIDIter := int64(bundleIDIter)
+		if bundle, ok := sc.bundles.Load(Key{sc.heightForFiringAuction, bundleIDIter}); ok {
+			bundle := bundle.(*Bundle)
+			fmt.Printf("BUNDLE ID: %d\n", bundleIDIter)
+
+			innerOrderMap := bundle.OrderedTxsMap
+			bundleSize := bundle.CurrSize
+			for bundleOrderIter := 0; bundleOrderIter < int(bundleSize); bundleOrderIter++ {
+				bundleOrderIter := int64(bundleOrderIter)
+
+				if scTx, ok := innerOrderMap.Load(bundleOrderIter); ok {
+					scTx := scTx.(*SidecarTx)
+					fmt.Printf("---> ORDER %d: %s\n", bundleOrderIter, scTx.Tx)
+				}
+			}
+		}
+	}
+	fmt.Println("-------------")
+}
+
+//--------------------------------------------------------------------------------
+
+// NOTE: not thread safe - should only be called once, on startup
+func (sc *CListPriorityTxSidecar) EnableTxsAvailable() {
+	sc.txsAvailable = make(chan struct{}, 1)
+}
+
+// Safe for concurrent use by multiple goroutines.
+func (sc *CListPriorityTxSidecar) TxsAvailable() <-chan struct{} {
+	return sc.txsAvailable
+}
+
+func (sc *CListPriorityTxSidecar) notifyTxsAvailable() {
+	if sc.Size() == 0 {
+		fmt.Println("[mev-tendermint]: ERROR: notified txs available but sidecar is empty!")
+	}
+	if sc.txsAvailable != nil && !sc.notifiedTxsAvailable {
+		// channel cap is 1, so this will send once
+		sc.notifiedTxsAvailable = true
+		select {
+		case sc.txsAvailable <- struct{}{}:
+		default:
+		}
+	}
+}
+
+//--------------------------------------------------------------------------------
+
+func (sc *CListPriorityTxSidecar) AddTx(tx types.Tx, txInfo TxInfo) error {
+
+	sc.updateMtx.RLock()
+	// use defer to unlock mutex because application (*local client*) might panic
+	defer sc.updateMtx.RUnlock()
+
+	// don't add any txs already in cache
+	if !sc.cache.Push(tx) {
+		sc.logger.Debug(
+			"rejected sidecarTx",
+			"reason", "already in cache",
+			"bundleId", txInfo.BundleID,
+			"bundleOrder", txInfo.BundleOrder,
+			"tx", tx.Hash(),
+		)
+		// Record a new sender for a tx we've already seen.
+		// Note it's possible a tx is still in the cache but no longer in the mempool
+		// (eg. after committing a block, txs are removed from mempool but not cache),
```

Comment:
I do not think this is true, you reset the cache in `Update`.

Reply:
Listed in https://github.com/informalsystems/audit-skip/issues/6

Reply:
Updated this comment in https://github.com/skip-mev/mev-tendermint/pull/131